migrations/2024-12-17-144959_community-post-tags/up.sql

```diff
@@ -0,0 +1,26 @@
+-- a tag for a post, valid in a community. created by mods of a community
+CREATE TABLE tag (
+    id serial PRIMARY KEY,
+    ap_id text NOT NULL UNIQUE,
+    name text NOT NULL,
+    published timestamptz NOT NULL DEFAULT now(),
+    updated timestamptz,
+    deleted timestamptz
+);
+
+-- indicates this tag was created by the mod of a community and can be applied to posts in this community
+CREATE TABLE community_post_tag (
+    community_id int NOT NULL REFERENCES community (id) ON UPDATE CASCADE ON DELETE CASCADE,
+    tag_id int NOT NULL REFERENCES tag (id) ON UPDATE CASCADE ON DELETE CASCADE,
+    published timestamptz NOT NULL DEFAULT now(),
+    PRIMARY KEY (community_id, tag_id)
+);
```

Comment:
It seems like you could get rid of this table and instead use a field `tag.community_id`, because each tag belongs to exactly one community.

Reply:
Didnt see that comment, but it goes directly against the RFC which specifies that tag urls look like this:
`"id": "community-url/tag/tag-id", // e.g. https://example.org/c/example/tag/News`

Which means that each tag belongs to a specific community. And if tags are global, it means that `lemmy.ml/tag/nsfw` and `lemm.ee/tag/nsfw` would use the same db entry. This would lead to all kinds of complications if one of the tags gets renamed or deleted. It also means that community mods cannot make any changes to tags as that could affect other communities or instances. I also dont like to add logic which "might be used later in the future", because most of the time that doesnt actually happen and so its a lot of complication for no purpose.

So tags should definitely be separate for each community, as specified in the RFC.

cc @dessalines

Reply:
I don't think it's a conflict. The tags added here are still all community-specific, we just store all the data for them in the generic `tags` table which we can in the future also use for something else.

The URL will still be as described in the RFC, the `tags` table will just be a collection of _all_ tags, even the ones not associated with a community (in the future). 

Whether or not they are community post tags is defined by whether an association (`tag_id`, `community_id`) exists in the `community_post_tags` table.

Reply:
The problem is that it adds a lot of extra complexity, as you already noticed with the db queries. This will get even worse when you start to work on federation, as you need to ensure that a mod in community A can make changes to tag `news`, without affecting the `news` tag in community B on a different instances. All this for no real benefit as its not clear if or when different types of tags will actually be supported. You should keep the implementation as simple as possible, otherwise it will become impossible to maintain.

Reply:
Post tags should be implemented based on the [RFC](https://github.com/LemmyNet/rfcs/blob/main/0004-post-tags.md), which specifically focuses on post tags to keep the initial implementation simple. If you want to do it differently and support all types of tags from the beginning, you need to start over and write a new RFC. Because I cant visualize how any of this will work without a much more detailed description.

Reply:
the complexity of the tags subquery in the posts view is mostly unrelated to whether we use a tags table or something more specific - if you look at the state before the changes i made for dess' comments, it was basically the same. I also don't expect too large differences in the rest of the implementation that I'll start with after this is merged.

it doesn't really add any complexity right now, but it's unclear if this is the right place of abstraction - we do not yet know exactly how other tag types might look, so we can not (and should not) perfectly plan for them. 

in any case, personally, i don't really care, so you two give me a decision. I'm fine with either how I added it initially or the changes I made. I don't think it makes a large difference right now, and this specifically we can still migrate later in either direction.

It's also not a large divergence from the RFC, though I do agree we should follow the RFC unless _new_ reasons not to come up.

Reply:
Get rid of the community_post_tag table and add a tag.community_id column. That column can be made nullable in the future.

Tag tables should definitely not be separated based on whether on not the tags are for a specific community. But I'm not sure if they should be separated based on what they are for (post, comment, user, etc). The advantage of that separation is it's simpler to restrict which tags can be added to something. The disadvantage is the database schema will be more repetitive, which is bad for maintainability.

Reply:
that would then be the same as my initial implementation, except having the table named differently